### PR TITLE
Add `.#{ime}` and `#{number}` syntax

### PR DESCRIPTION
**Pull request by NduatiK**
Depends on https://github.com/liveview-native/live_view_native_stylesheet/pull/34.

Adds support for:
```swift
  "button-" <> style do
    buttonStyle(.#{style})
  end

  "h-" <> height do
    height(#{height})
  end
```

Which lets us do:
```elixir
<Button class="h-16 button-bordered"> ... </Button>
```

### CODE REVIEW

**Comment by bcardarella** [comment]
I'd like to get @carson-katri 's thoughts on this

**Review by carson-katri — COMMENTED** [review]
Would it be possible to have it do simple string interpolation instead of being locked into specific use cases (`to_ime` and `to_number`)? Then it could be used as a more general interpolation feature. Or is it more secure to lock into a few specific cases?

File: lib/live_view_native/swiftui/rules_parser/modifiers.ex [hunk 1]
```diff
@@ -58,21 +58,21 @@ defmodule LiveViewNative.SwiftUI.RulesParser.Modifiers do
...
-      empty()
-    else
-      ignore(string("."))
-    end
-    |> ignore(string("to_ime"))
+    # if is_initial do
+    #   empty()
+    # else
+    #   ignore(string("."))
+    # end
```
[the remark this answers is not in the document]
Comment (NduatiK) [reply lib/live_view_native/swiftui/rules_parser/modifiers.ex:65]: > Can this be removed
Yes! Thank you.

---

> Would it be possible to have it do simple string interpolation ...

TLDR: I think the current approach is the best given the parser implementation we have, but I'm open to what else we should support.

It sounds like you want to be able to do the normal string interpolation so that something like the following could work in future:

```elixir
# the style
"h-" <> height do
  height(#{height})
end

# Using the class
<Button class="h-:infinity"> ... </Button>

#expected output
height(:infinity)
```

Unfortunately, I don't think that would work. The current model assumes that the variable part of the class name is known at runtime (_after_ we have already compiled the AST).

```mermaid
flowchart LR
    A(SHEET) -->|Compile Time| B(class function)
    B         -->|Runtime| C(AST)
    :infinity -->|Runtime| C(AST)
```

The purpose of the helper functions (`to_ime`, `to_float` and `to_integer`) was to allow us to map runtime strings into other types.

Let me know if you have any specific use cases, you'd have wanted to support, and we can extend this PR

[the remark this answers is not in the document]
Comment (carson-katri) [reply lib/live_view_native/swiftui/rules_parser/modifiers.ex]: Interesting, I didn't realize those were evaluated at runtime. Thanks for the clarification.

[the remark this answers is not in the document]
Comment (bcardarella) [reply lib/live_view_native/swiftui/rules_parser/modifiers.ex]: @carson-katri we're going to no use this syntax as we want to to reserve the Elixir interpolation for instances of dropping down into writing the functions by hand. We'll do what Heex has and use `{ ...  }` instead for interpolation within the sheet.

As far as executing Eixir functions, I don't think that's a good idea for us to do in the sheet because of the `attr` accessor. That value isn't accessible at compile-time or runtime but only when yanked from the element by the SwiftUI client.

In your examples above the recommendation for now is to write a custom `class/2` function if you more complex manipulations:

```elixir
def class("frame-" <> dims, _target) do
   [width, height] = String.split(dim, ":")

   ~RULES"""
   frame(width: #{width}, height: #{height}
   """
end
```

**Comment by bcardarella** [comment]
Reopening as @NduatiK will update the syntax to `{...}`

This shoudl be such that wraping in the moustache syntax is now a requirement for interpolating values into the AST. For example:

```
"foobar-" <> foobar do
   barbaz(foobar)
end
```

should no longer be valid. Instead it must be:

```
"foobar-" <> foobar do
   barbaz({foobar})
end
```

so if the class match is "red" it would then be parsed as if the value is injected directly into the rules as:

```
barbaz(red)
```

Any type of changes to the value will require an external wrap, for example for strings:

```
"foobar-" <> foobar do
   barbaz("{foobar}")
end
```

will now result in:

```
barbaz("red")
```

and to cover to the dot-notation:

```
"foobar-" <> foobar do
   barbaz(.{foobar})
end
```

gives us

```
barbaz(.red)
```

and this will then be parsed into our AST node.

**Comment by bcardarella** [comment]
Small correction, the following would be valid and should compile:

```
"foobar-" <> foobar do
   barbaz(foobar)
end
```

but it would compile to:

```
barbaz(foobar)
```

**Comment by carson-katri** [comment]
I think requiring `"{variable}"` adds too much friction for simple use cases. I think most developers would expect this to resolve to `modifier("whatever the value is")`:

```elixir
"my-class-" <> variable do
  modifier(variable)
end
```

Then the `{...}` syntax lets you do something else with that string variable.

**Comment by carson-katri** [comment]
That would be similar to how I can use variables from `~RULES`:

```elixir
variable = "test"
~RULES"""
modifier(variable)
"""
```
I would expect the value of `variable` to be used here, not the name of the variable.

**Comment by NduatiK** [comment]
Sounds like the only change is to support `.{variable}`. 

---

> Small correction, the following would be valid and should compile:
> 
> ```
> "foobar-" <> foobar do
>    barbaz(foobar)
> end
> ```
> 
> but it would compile to:
> 
> ```
> barbaz(foobar)
> ```


By default, variables from the block head are strings.
```elixir
"foobar-" <> foobar do
   barbaz(foobar)
end
```

will inject the value of `foobar` into the AST as a string so there's no need to separate `foobar` and `"{foobar}"`.

**Comment by bcardarella** [comment]
@carson-katri what we want to do is move closer to what the HEEX templates provide within our own template. This would require us to use interpolation for any value injection into the template. Otherwise the text should be treated as a literal and not interpolated.

**Comment by NduatiK** [comment]
@bcardarella, could we avoid faking interpolation for now? It introduces quite a bit of complexity for something we are planning to change soon.

---

If this is something we must have, we need to consider how we handle integers and floats

```elixir
"foobar-" <> foobar do
   barbaz({to_float(foobar)})
      # OR
   barbaz({foobar})
end
```

**Comment by carson-katri** [comment]
@bcardarella I don’t think these need to behave exactly the same. In a HEEx template, you’re interpolating values that exist outside the normal markup syntax. If the variable declaration happened in a separate nested language (like nesting elixir in markup), then interpolation would make sense. But here, the variable is declared within the stylesheet, so I would expect it could be used directly.

**Comment by bcardarella** [comment]
@carson-katri right but that's exactly what is happening here. The `~SHEET` sigil of:

```
"color-" <> color do
   color(.{color})
end
```

compiles to:

```elixir
def class("color-" <> color, _target) do
   ~RULES"""
   color(.{color})
   """
end
```

The `~SHEET` sigil is just a convenience DSL

**Comment by bcardarella** [comment]
@NduatiK putting the literal into the document means we don't need a way to typecast. The parsing would happen on the text after the interpolation.

```
<Text class="size-12">Hello</Text>
```

```
"size-" <> size do
   fontSize({size})
end
```

after interpolation the rule will look like:

```
fontSize(12)
```

this should hold up for floats as well:

```
<Text class="size-12.4">Hello</Text>
```

```
fontSize(12.4)
```

**Comment by NduatiK** [comment]
Implementing real interpolation is not easy. For the v0.2 release, I can implement a parser that convert the following:

|  Stylesheet | Result  |
| ---  |  ---  |
| `"{ variable }"` | `ast(variable)`  |
| `{ variable }` | `ast(to_number(variable))`  |
| `.{ variable }` | `ast(to_ime(variable))`  |

to_number is implemented as shown [here](https://github.com/NduatiK/liveview-client-swiftui/blob/4c6374cd9347368e891a36dcbe9516dbe4cf8594/test/mocks/mock_sheet.ex#L30)

**Comment by NduatiK** [comment]
@bcardarella, any feedback on this?

**Comment by bcardarella** [comment]
@NduatiK I plan on circling back this week hopefully after I've completed my refactor for `render_with` integration. Sorry for the delay 😞

**Comment by NduatiK** [comment]
No problem, just wanted to make sure we aren't both waiting on each other🙃

**Comment by bcardarella** [comment]
I believe this is no longer valid given https://github.com/liveview-native/live_view_native_stylesheet/pull/42
